### ISSUE TITLE
Hierarchical Classes Bounded/Unbounded Random numbers

### ISSUE THREAD

**Issue by lehins** [lehins/random#8]
This discussion has already popped up thanks to @Shimuuar
> uniformR says generate all possible values with equal probability. Integer works just fine with this. But uniform? No way! There's infinite number of integers

It would make absolute sense to split number generation into two(three) groups:
* Bounded: `Int`, `Word`, etc:
* Unbounded: `Integer`, `Natural`
  * Potentially also: `Float` and `Double` (`CFloat` and `CDouble` too of course)
    Reasons I consider including floating point numbers here instead of with fixed width integral types because: 
      * There are whole lot of binary representations of floating numbers that aren't really valid: `NaN`,`+/-0` and `+/-Infinity`
      * Floating point numbers don't have `Bounded` instance

What are our options:

1. Leave it broken as it is now in `Random` and in `Variate` classes.

2. Introduce `RandomR` for all types that can be generated, while using `Random` only for things that has bounds:

```haskell
class RandomR a where
  randomRM :: MonadRandom g m => (a, a) -> g -> m a
  randomR :: NoSplitGen g => (a, a) -> g -> (a, g)
  randomR r g = runStateGen g (genRandomR r)

class (Bounded a, RandomR a) => Random a where
  randomM :: MonadRandom g m => g -> m a
  randomM = randomRM (minBound, maxBound)
  random  :: NoSplitGen g => g -> (a, g)
  random g = runStateGen g genRandom
```

This will be a breaking change for users that are relying on `random` to generate floating point numbers and unbounded integral types.
Conceptually though, I think this is a good distinction to have.

Thoughts?

**Comment by Shimuuar** [comment]
I think that neither `Bounded` nor `RandomR` should be superclass of `Random`. For example `(Word8,Word8)` is perfectly fine instance for `Random` (or any product type really). But those don't necessarily have sensible `RandomR` nor are `Bounded`

**Comment by lehins** [comment]
`Bounded` enforces that the type indeed can have a uniform distribution. Nothing prevents anyone create an orphan instance for `Integer` for example.

I don't think product types should have instances of `Random`. Precisely because this hierarchy gets broken and anyone can do `(,,) <$> random <*> ranomdR` if they need random tuples.

Also this is terrible ux `foo :: (Random r, RandomR r) => .. -> r`

There is a natural hierarchy for random numbers (which is what we care about), therefore I am really against `RandomR` not being a superclass of `Random`. I would honestly prefer having a single class `Random` rather than not have this hierarchy.

**Comment by Shimuuar** [comment]
> `Bounded` enforces that the type indeed can have a uniform distribution. 

In fact no. It only assert that set of values has total order and both maximum and minimal value. And one could have both finite set without order: take for example any finite group. It's finite, `Random` is well defined there, yet there's no sensible order. And value could be both `Ord` and `Bounded` and still infiinite. Here is contrived example

`X` represents numbers of form `1/n`

```haskell
newtype X  = X Integer
instance Ord X where
  ...
instance Bounded X where
  minBound = X (-1)
  maxBound = X 1
```

> Also this is terrible ux `foo :: (Random r, RandomR r) => .. -> r`

`type RandomBoth r = (Random r, RandomR r)` to the rescue! (Names up to debate, of course)

Also haskell history show that there's clear trend for splitting type classes. Semigroup was split from Monoid. Num & Co is not split only because it would be way too painful. I think we shouldn't repeat mistakes and bundle together operations when only one or other could be defined for some types.

**Comment by lehins** [comment]
With `Bounded` I don't feel too strongly about, so that isn't a problem.

I do think a split into two classes is a good idea, but just like it is with `Semigroup` and `Monoid` they have hierarchy.

You are right, we should not repeat mistakes, just look at FAM proposal.

This doesn't solve anything: `type RandomBoth r = (Random r, RandomR r)` type inference doesn't care about those type synonyms. I learned while back that in public APIs type synonyms only bring confusion and don't solve anything. `conduit` is a really good example of that.

**Comment by Shimuuar** [comment]
> I do think a split into two classes is a good idea, but just like it is with Semigroup and Monoid they have hierarchy.

But my point is that they don't. `UniformR` imply existence of order but `Uniform` doesn't care about order. It's very easy to construct realistic examples (see above) for types that could have instance for one type class but not another

**Comment by lehins** [comment]
I understand your point with `X` type and I understand that we can come up with bunch of different types that don't follow this hierarchy, hell even product types, as you mention, don't.

So the question is should they have Random/RandomR instances at all? I feel we need to keep `Random` for primitive like types, not for any possible type out there.

Take a list for example, we can construct an instance for `Random a => Random [a]`, by creating a random length and then random elements, but that's not we want for `random` package.

If someone has an esoteric type they can use `random` to construct it, but `Random` instances will not be polluted with such type. For public APIs exporting helper functions that generate such types is a better way to go, for example: `randomList :: Random a => Int -> [a]` and `randomListN :: RandomR a => Int -> (a, a) -> [a]` are perfectly fine functions.

**Comment by Shimuuar** [comment]
> I feel we need to keep Random for primitive like types, not for any possible type out there.

I think this is difference in our approaches. I try to define laws for type class after that every type that have unique instance that satisfy laws should have it defined. It's ability to define type classes for user defined types makes them that useful and laws make it possible to use type classes without thinking about concrete implementation

I took law for `Random` is to generate every inhabitant of type with equal probability. This law immediately rules out instance for lists btw. It have infinite number of inhabitants.

**Comment by lehins** [comment]
@Shimuuar I absolutely agree with that.

> I think this is difference in our approaches.

I am trying to be less of a perfectionist and a theorist, I want to be practical. For me usability and performance are priorities.

If there is a need for such type classes I think they should be in a separate library. Possibly a library that generates all other distributions and for all sort of other types.
> It's ability to define type classes for user defined types makes them that useful and laws make it possible to use type classes without thinking about concrete implementation

The goal for `random` should be a practical building block for other random libraries to use that get a unified interface to available RNGs and generation of basic types. Which is exactly what it currently is, except it sucks at performance and interface is too outdated.

**Comment by Shimuuar** [comment]
But type classes which could be used by other packages _are_ practical. Having two type classes in two libraries for doing same thing is very impractical thing to have. 

At this point I think we made our position clear. Maybe someone else has something to say

**Comment by lehins** [comment]
Keep in mind that without having `RandomR a => Random a` it would a very breaking change that will affect *all* calls to `randomR` and `randomRs` out there in the wild. To give you a concrete example `choose` from quickcheck has not been changed for 14 years: 
https://github.com/nick8325/quickcheck/blame/master/Test/QuickCheck/Gen.hs#L136

Which means everybody using `choose` has to change their type signature as well. That is a lot of code and I really want to avoid. Especially not because there is some theoretical improvement that we can make.

**Comment by Shimuuar** [comment]
Maybe better idea is to split `Random` into three type classes:

```haskell
class Uniform a where
  uniform :: MonadRandom g m => g -> m a

class UniformRange a where
  uniform :: MonadRandom g m => g -> (a,a) -> m a

class (Uniform a, UniformRange a) => Random
```

This way we get type classes with sensible semantics. Another possibility is to introduce type classes above and leave `Random` as it is and deprecate it slowly. This is probably best approach. It induces least amount of breakage, and there's clear migration path

Wrt original proposal  I think that introducing breaking change to type class and leaving it broken just doesn't make sense. And proposed change is breaking. Float, Double, Integer are no longer have Random instance. All instances for custom types are broken too. Also `Random` mixes so much in single type class that it's not possible to fix it without massive breakage

**Comment by lehins** [comment]
@Shimuuar There are a couple of minor problems with this idea, but I think I see where you are going with it. Problems are:
* I guess this is just a typo: `UniformRange` should have a function named `uniformR` 
* `(Uniform a, UniformRange a)` should not be superclasses of `Random`, but it can use those instances as default implemntation.

I am gonna code it up real quick in a separate PR. I let 'yall know when I am done

**Comment by Shimuuar** [comment]
>    I guess this is just a typo: UniformRange should have a function named uniformR
Yep. Copy paste typo.

>    (Uniform a, UniformRange a) should not be superclasses of Random, 
I started with idea of adding them as superclasses to `Random` but while writing reply come to comclusion that it's probably better to don't change Random at all. Excluding maybe default method definitions

>  but it can use those instances as default implemntation.
You mean `DefaultSignatures`? Yes that will work

**Comment by lehins** [comment]
@Shimuuar 
> You mean DefaultSignatures?

That's exactly what I meant.

>  it's probably better to don't change Random at all. 

I sort of agree with this. I understand that you think that `Random` is broken from mathematical perspective, but I don't think we should completely deprecate `Random`

I just looked at how Rust does it (thanks to @curiousleo for links in #13) and they have a [`Standard` distribution](https://rust-random.github.io/rand/rand/distributions/index.html#the-standard-distribution) that has no mathematical significance but is used purely for generating "default" random values for various data types. 

If we leave `Random` around for the same purpose (which is mostly what it is used for now), then we could in the future add instances for more complex types like product and sum types. This would mean only `random` and `randomM` functions would survive, because `randomR` and others would not make sense.

Note that with `Uniform` and `UniformRange` you can't have instances for complex types that intermix those two classes. So for examples you can only have:

```haskell
instance (Uniform a, Uniform b) => Uniform (a, b)

instance (UniformRange a, UniformRange b) => UniformRange (a, b)
```

`Random` could be a decent middle ground for such cases, because we would still have instances for infinite types like `Random Integer`.

**Comment by lehins** [comment]
@Shimuuar See #14 for a potential implementation of your suggestion. Let me know what you think.

I personally really like what you suggested, especially when mixed with using some "default" ranges for particular types in `Random`. I think this way we can have the best of both worlds: 

* Correct uniform distributions: useful for math
* ability to generate generic "random" for diverse collections of types: useful for programmers

**Comment by Shimuuar** [comment]
RE deprecation of `Random`. What I have in mind is very slow and gradual deprecation. Start by adding note that use of this type class API is discouraged, provide migration path. Don't add DEPRECATED pragma yet. Just call it legacy API that would be enough to deter programmers. Then wait for a few years. If user of API do start transitioning to new style API and it is actually used DEPRECATED could be added.

> ability to generate generic "random" for diverse collections of types: useful for programmers

I would say that it's useful as long as result type is known at call site. Then it's possible to reason about what generator actually do. Once you have `∀a. Random a => m a` it becomes much more difficult.

**Comment by lehins** [comment]
Not at all. It is not slightest bit harder than `∀a. Num a => fromInteger :: Integer -> a` 
> Once you have ∀a. Random a => m a it becomes much more difficult.

Every instance of `Random` can be easily documented since it will be doing a single thing.  For example:

```haskell
-- | Generates a 64bit floating point value in a range @[0, 1)@
instance Random Double where
```
(Or whatever else range we decide in #6)

With such haddock a user at a call site can easily reason what the value will be, because eventually a user will have to decide on the type being generated.

Again, I would not want to deprecate `Random` class completely, just most (but not all) functions within it. 

Why shouldn't we add `DEPRECATED`? This is the best way to let everyone know about the actual deprecation. A few days ago you were ok with breaking all call sites of `randomR`, but now a compile time warning is too much? :smile:  
> Don't add DEPRECATED pragma yet.

**Comment by curiousleo** [comment]
```
(UniformRange a, Bounded a) => Uniform a
```

**Comment by Shimuuar** [comment]
> Not at all. It is not slightest bit harder than ∀a. Num a => fromInteger :: Integer -> a

Objection! With `Num` there's implied expectation that `fromInteger` commutes with other operations from `Num`: `fromInteger a + fromInteger b = fromInteger (a + b)`. This even holds for floating point! (Approximately as all floating point things)

As I said it's very much possible to  use API like `Random` when you know which instance you select. Problem is you don't when you write code like:
```haskell
foo :: Random a => ... -> m a
```
or
```haskell
instance Random a => Random (Pair a) where
  ...
```
It could use any instance and without laws it could be difficult to understand what such code will do

> A few days ago you were ok with breaking all call sites of randomR, but now a compile time warning is too much? smile

`That's very simple. I  think that it's very important to create good type classes to build rest of ecosystem on top. If one have to break Random to do so, well what have to be done have to be done. If we create entirely new API we could wait with deprecation for some time and check whether new API has any problem in real world use which should be addressed. This could require breaking change.

**Comment by lehins** [comment]
I am glad you bring up the example of `fromInteger a + fromInteger b = fromInteger (a + b)`, because `(+)` is part of `Num`, and there is indeed a mathematical law for it. But `Random` would not have any functions associated with it and if you can't come up with laws for it it does not mean it is not useful. Hell, so many languages have the same concept and you are suggesting they are all wrong? 

My argument about `fromInteger` was being exactly the same as `random` on it's own. Do you know what the value of `Num a => fromInteger 277 :: a` will be until you know the type of `a`? You don't! 

It could be `277` or it could be `21`

Same with `random`, it might be hard to "understand what such code will do", but if it is hard, don't write it. Restrict the function to your own types, or create custom classes with laws that use `Random` class underneath if you wish. But as far as `Random` is concerned, in my mind, the law is: "Document what your instance generates for your particular type!"

I'll give you a good example. `Arbitrary` from `QuickCheck`. Are there any laws for it? Can you argue about what sort of values the generator will produce without looking through the code and/or documentation. Probably not, I certainly can't. But that type class is incredibly useful. I can find many such examples in the wild.

I guess my final argument is: if the concept comes from some mathematical concept (eg. Uniform distribution, associativity, addition, etc.....) it comes with laws. `Random` doesn't really come from math. And it's current instances certainly don't follow [m]any laws :)

**Comment by lehins** [comment]
@Shimuuar I am sure you know this, but I forgot to mention it nevertheless that even for `Num` there are laws that have to be broken in order to be practical: eg. `Num a => abs (fromInteger x) >= (0 :: a)` for example doesn't hold for `x = toInteger (minBound :: a)`

**Comment by lehins** [comment]
@Shimuuar You know, I think the best way I can show how `Random` with only two functions `random` and `randomM` can be useful, even without any laws, is on the example of [`Default`](https://hackage.haskell.org/package/data-default-class-0.1.2.0/docs/Data-Default-Class.html#t:Default) class. There are no laws for it and you can't really know what value `def` will produce without looking at the implementation for a particular type, nevertheless this class is tremendously useful and there are thousands of packages that depend on it, if not directly then transitively.

I am pretty sure there are types out there that can't have an instance of `Uniform` or `UniformRange` due to some specific issue with that type, in which case users can still make use of `Random`. A good example would be a type that encodes a Social Security number in the US. there are numbers that are no longer valid, there are ranges that have been reserved. So you can't truly generate a uniform distribution of all SSNs, nevertheless you can take as much knowledge you have about SSN and create a `Random` instance for such type.

In short, it is not all about math, we often have to deal with real world when we write programs.

**Comment by Shimuuar** [comment]
I think we're moving too deep into philosophy territory. :-) I won't comment further

P.S. As for `abs` it's broken part of very broken haskell's numeric tower.

**Comment by lehins** [comment]
Concur
>  As for abs it's broken part of very broken haskell's numeric tower.

**Comment by idontgetoutmuch** [comment]
@Shimuuar @lehins I think we have implemented this so can we close this issue?

**Comment by lehins** [comment]
That's right. There is still a question of `Random` class, but one way or another for now we have to keep it for backwards compatibility.

One question I would like to ask you guys is how would you fee about renaming:
* `uniform` -> `uniformM`
* `uniformR` -> `uniformRM`

Just so we can see their monadic nature in the name. Another reason I am asking is because we currently don't have matching pure functions that do the same, both of which have trivial implementation:

```haskell
uniform :: (RandomGen g, Uniform a) => g -> (a, g)
uniform g = runGenState g uniformM

uniformR :: (RandomGen g, UniformRange a) => (a, a) -> g -> (a, g)
uniformR r g = runGenState g (uniformRM r)
```

**Comment by curiousleo** [comment]
I'm in favour of the suggested renaming and of adding the pure `uniform` and `uniformR` functions.

**Comment by idontgetoutmuch** [comment]
Good idea

**Comment by curiousleo** [comment]
This issue has been inactive for a while, and I think the discussion has reached a conclusion. Closing.

We now have:
- [pure `uniform`](https://htmlpreview.github.io/?https://raw.githubusercontent.com/idontgetoutmuch/random/haddock-preview/docs/System-Random.html#v:uniform)
- [pure `uniformR`](https://htmlpreview.github.io/?https://raw.githubusercontent.com/idontgetoutmuch/random/haddock-preview/docs/System-Random.html#v:uniformR)
- [the `Uniform` type class](https://htmlpreview.github.io/?https://raw.githubusercontent.com/idontgetoutmuch/random/haddock-preview/docs/System-Random.html#t:Uniform)
- [the `UniformRange` type class](https://htmlpreview.github.io/?https://raw.githubusercontent.com/idontgetoutmuch/random/haddock-preview/docs/System-Random.html#t:UniformRange)
